crates/plugins/src/git.rs

```diff
@@ -0,0 +1,71 @@
+use anyhow::{anyhow, Result};
+use std::path::{Path, PathBuf};
+use tokio::process::Command;
+use url::Url;
+
+const DEFAULT_BRANCH: &str = "main";
+
+/// Enables cloning and fetching the latest of a git repository to a local
+/// directory.
+pub struct GitSource {
+    /// Address to remote git repository.
+    source_url: Url,
+    /// Branch to clone/fetch.
+    branch: String,
+    /// Destination to clone repository into.
+    git_root: PathBuf,
+}
+
+impl GitSource {
+    /// Creates a new git source
+    pub fn new(
+        source_url: &Url,
+        branch: Option<String>,
+        git_root: impl AsRef<Path>,
+    ) -> Result<GitSource> {
+        Ok(Self {
+            source_url: source_url.clone(),
+            branch: branch.unwrap_or_else(|| DEFAULT_BRANCH.to_owned()),
+            git_root: git_root.as_ref().to_owned(),
+        })
+    }
+
+    /// Clones a contents of a git repository to a local directory
+    pub async fn clone(&self) -> Result<()> {
+        let mut git = Command::new("git");
+        git.args([
+            "clone",
+            self.source_url.as_ref(),
+            "--branch",
+            &self.branch,
+            "--single-branch",
+            &self.git_root.to_string_lossy(),
+        ]);
```

Comment:
Thanks for pointing this out! This is because some inportant information may be dropped in the lossy conversion, correct? So data used by functions shouldnt undergo this but if its for logging information, it isnt risky?